core/src/main/java/org/apache/calcite/rel/type/RelDataTypeFactoryImpl.java

```diff
@@ -612,8 +612,15 @@ public SqlCollation getCollation() {
     }
 
     public SqlTypeName getSqlTypeName() {
+      Class clz =  clazz;
```

Comment:
I think this logic would be more appropriate inside the JavaToSqlTypeConversionRules.

Reply:
Yes, you are right.

Reply:
+1

Reply:
Currently, there's something wrong for checking subclass, so I remove the logic of checking subclass.
I'll create another pr after I figured out the reason.